src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java

```diff
@@ -0,0 +1,475 @@
+    builder.setBuildId(buildId.toString());
+    builder.setInstanceName(instanceName);
+    builder.setDigestFunction(digestFunction);
+    builder.setOutputPathPrefix(outputPathPrefix.toString());
+    Path originalOutputPath = execRoot.getRelative(relativeOutputPath);
+    builder.putOutputPathAliases(originalOutputPath.toString(), ".");
+    StartBuildResponse response = blockingStub.startBuild(builder.build());
+
+    // Replace the output path with a symbolic link pointing to the
+    // directory managed by the remote output service.
+    PathFragment outputPath = outputPathPrefix.getRelative(response.getOutputPathSuffix());
+    try {
+      try {
+        originalOutputPath.deleteTree();
+      } catch (FileNotFoundException e) {}
+      originalOutputPath.createSymbolicLink(outputPath);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+
+    currentBuildId = buildId;
+    currentOriginalOutputPath = originalOutputPath.asFragment();
+    currentRelativeOutputPath = PathFragment.create(relativeOutputPath);
+
+    if (previousBuildId == null || !response.hasInitialOutputPathContents()) {
+      // Either Bazel or the remote output service has performed no
+      // build before.
+      return ModifiedFileSet.EVERYTHING_MODIFIED;
+    }
+    InitialOutputPathContents initialContents = response.getInitialOutputPathContents();
+    if (initialContents.getBuildId() != previousBuildId.toString()) {
+      // Bazel and the remote output service disagree on the build ID of
+      // the previous build.
+      return ModifiedFileSet.EVERYTHING_MODIFIED;
+    }
+    // Bazel and the remote output service agree on the build ID of the
+    // previous build. Return the set of paths that have been modified.
+    //
+    // TODO: Do these paths need to be relative to the exec root or the
+    // output path? Assume the exec root for now.
+    return ModifiedFileSet.builder()
+        .modifyAll(
+            Iterables.transform(
+                initialContents.getModifiedPathsList(),
+                (p) -> currentRelativeOutputPath.getRelative(p)))
+        .build();
+  }
+
+  @Override
+  public void finalizeBuild(boolean buildSuccessful)
+      throws BuildFailedException, AbruptExitException, InterruptedException {
+    FinalizeBuildRequest.Builder builder = FinalizeBuildRequest.newBuilder();
+    builder.setBuildId(currentBuildId.toString());
+    builder.setBuildSuccessful(buildSuccessful);
+
+    blockingStub.finalizeBuild(builder.build());
+
+    previousBuildId = currentBuildId;
+    currentBuildId = null;
+    currentOriginalOutputPath = null;
+    currentRelativeOutputPath = null;
+  }
+
+  @Override
+  public void finalizeAction(Action action, MetadataHandler metadataHandler)
+      throws IOException, EnvironmentalExecException, InterruptedException {
+    // TODO: Would this be the right place to call into the remote
+    // output service to check whether any I/O errors occurred? If so,
```

Comment:
The FUSE file system may be lazy loading, meaning that I/O errors, transient or permanent, may occur while a locally running build action attempts to access input files that were generated by a remote action. Right now these errors will get logged by the FUSE daemon, but the system call that triggered the error will return EIO.

Ideally, the FUSE daemon would record the errors instead of logging them, so that Bazel can fetch it after a locally running build action terminates. That way Bazel can display a rich error message or retry executing the action.

This is something that Buildbarn already implements at the worker level. I/O errors in the FUSE file system will get attached to the ExecuteResponse.

Reply:
That's a very interesting idea, thanks for explaining it.

Reply:
I am sorry, I think I may need a little more explanation on this one. To me, we are talking about 2 cases here:

1. Errors which happen at the time of running an action remotely when we inform the FUSE/NFS about incoming files. If that does anything which can fail, I would think that the API responses should be a place to report that.
2. Error which happen after the action is done and a dependency of it tries to read the files locally -- in this case, the regular FS-level errors should suffice (that I consider no different from user trying to run `less bazel-out/file` and that failing).

On the surface, that seems to me that we should never need this call for the purpose mentioned in the TODO, which is why I think I must be missing something. Would you mind adding a little more detail, please?